### ISSUE TITLE
`isValidPath()` logic does not match actual implementation

### ISSUE THREAD

**Issue by clayton-cc** [magnetikonline/edgy#2]
Hello, I've started playing with the library to set up full tests of our Lambda@Edge functions. Ideally, I'd like to chain together the 4 functions to replicate and test our flow in an offline context. We use Lambda@Edge for split testing across our sites so we modify and overwrite some of the event values.

When testing the `edgy.OriginRequest`, we noticed the `isValidPath` logic does not match the real implementation, and the AWS documentation is a bit misleading. In the [Lambda Edge event structure docs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html) on `path` it says:

> The directory path at the origin where the request should locate content. The path should start with a forward slash (/) but shouldn't end with one

However what is missing from this that is in the [Cloudfront Developer Guide](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-web-values-specify.html#DownloadDistValuesOriginPath). Where it starts with:

> If you want CloudFront to request your content from a directory in your origin

Therefore I believe that the default value for telling cloudfront _not_ to grab content from a different directory is just an empty string, instead of `/`. I tested this with a live cloudfront distribution, and if you pass a `/`, it will raise an error while an empty path or a "valid" path (something like `/<env>/<version>`) will work as expected. 

I have updated this function locally to continue working through implementing testing and will hold any PRs in case of other fixes

**Comment by magnetikonline** [comment]
Hello @clayton-cc - thanks for reaching out with the issue.

Just had to refresh myself with my own code 😄. This is interesting and yes, I took the description outlined in the event structure to read that `"path"` must be either `/` or a `/path/to/resource`.

On re-read, it seems I may very well have that wrong - and in addition the example given on the event structure page itself [for an example origin request](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html#example-origin-request) shows the following:

```json
"origin": {
  "custom": {
    "customHeaders": {},
    "domainName": "example.org",
    "keepaliveTimeout": 5,
    "path": "",
    "port": 443,
    "protocol": "https",
    "readTimeout": 30,
    "sslProtocols": [
      "TLSv1",
      "TLSv1.1",
      "TLSv1.2"
    ]
  }
```

yes, an empty `"path"` 😄.

So I think the fixes should be:

- Update `isValidPath()` under `payloadVerifyRequestOrigin()` to allow either empty string, or a string starting with a `/` - but not ending with `/` (current behaviour). But it should _no longer_ allow just `/` as per your tests.
- Update `setOriginCustom()` and `setOriginS3()` methods to ensure the given `path` is also validated by `isValidPath()` logic - and throw an exception if not.

I'm going to find some time to do the same as you - spin up a test/POC CloudFront/S3 and Lambda@Edge function on the `origin request` point in the request lifecycle and confirm these settings of `path` first.

**Comment by magnetikonline** [comment]
Sorry @clayton-cc - haven't forgotten about this - just trying to find the time to test/remediate. 😄

**Comment by magnetikonline** [comment]
Working through this now - some prelim details:

Setting `request.origin.s3.path = '/';`:

```
The Lambda function returned an invalid origin configuration: The value of OriginPath is invalid. We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.
```

**Comment by magnetikonline** [comment]
Hello @clayton-cc - I'm _finally_ trying to get around to this issue - yes, it's been over a year!

Yep - based on my tests above with a CloudFront distribution, you're 100% correct - I've currently got the rules/tests slightly wrong ☝️.

Will now work to fix this / make better.

**Comment by magnetikonline** [comment]
Have completed the fixes here finally! 🎉

Addressed in commits: ce7516b3e0dba2303525b029f5bc6745cb0457ae, 6dc348ec4f6ee1fbd000f02667bf8767e93cdffa, 93a32515fc77139bd6048cdaa8680517f84e76d7.

All part of release `v1.3.0+`.

Thanks again @clayton-cc for bringing this issue to my attention 👍
